rholang/src/test/scala/coop/rchain/rholang/interpreter/MatchTest.scala

```diff
@@ -97,11 +98,88 @@ class VarMatcherSpec extends FlatSpec with Matchers {
     result should be(Some(Map[Int, Par](0 -> GPrivate("zero"), 1 -> GPrivate("one"))))
   }
+
+  "Matching a receive with a free variable in the channel and a free variable in the body" should "capture for both variables." in {
+    val target: Receive = Receive(
+      List(
+        ReceiveBind(List(Quote(EVar(FreeVar(0))), Quote(EVar(FreeVar(1)))), Quote(GInt(7))),
+        ReceiveBind(List(Quote(EVar(FreeVar(0))), Quote(EVar(FreeVar(1)))), Quote(GInt(8)))
+      ),
+      Send(Quote(GPrivate("unforgeable")), List(GInt(9), GInt(10)), false, 0, BitSet()),
+      false,
+      4,
+      2
+    )
+    val pattern: Receive = Receive(
+      List(
+        ReceiveBind(List(Quote(EVar(FreeVar(0))), Quote(EVar(FreeVar(1)))), Quote(GInt(7))),
+        ReceiveBind(List(Quote(EVar(FreeVar(0))), Quote(EVar(FreeVar(1)))), ChanVar(FreeVar(0)))
+      ),
+      EVar(FreeVar(1)),
+      false,
+      4,
+      2
+    )
+    val result = spatialMatch(target, pattern).runS(emptyMap)
+    result should be(
+      Some(
+        Map[Int, Par](
+          0 -> GInt(8),
+          1 -> Send(Quote(GPrivate("unforgeable")), List(GInt(9), GInt(10)), false, 0, BitSet()))))
+  }
+
+  "Matching an eval with no free variables" should "Succeed, but not capture anything." in {
+    val target: Eval  = Eval(ChanVar(BoundVar(0)))
+    val pattern: Eval = Eval(ChanVar(BoundVar(0)))
+    val result        = spatialMatch(target, pattern).runS(emptyMap)
+    result should be(Some(Map.empty[Int, Par]))
+  }
+
+  "Matching between New's" should "match the bodies if the new count is the same" in {
+    val target: New =
+      New(2,
+          Par()
+            .prepend(Send(Quote(GInt(7)), Seq(GInt(42)), false))
+            .prepend(Send(ChanVar(BoundVar(0)), Seq(GInt(43)), false, locallyFree = BitSet(0))))
+    val pattern: New =
+      New(2,
+          Par()
+            .prepend(Send(Quote(GInt(7)), Seq(EVar(FreeVar(0))), false, freeCount = 1))
+            .prepend(EVar(Wildcard(WildcardMsg()))))
+    val result = spatialMatch(target, pattern).runS(emptyMap)
+
+    result should be(Some(Map[Int, Par](0 -> GInt(42))))
+  }
+
+  "Matching between matches " should "require equality of cases, but match targets and inside bodies." in {
+    val target: Match =
+      Match(
+        EList(Seq(GInt(4), GInt(20))),
+        Seq(
+          MatchCase(EList(Seq(EVar(FreeVar(0)), EVar(FreeVar(1)))),
+                    Send(Quote(EVar(BoundVar(1))),
+                         Seq[Par](EVar(BoundVar(0))),
+                         false,
+                         locallyFree = BitSet(0, 1))),
+          MatchCase(EVar(Wildcard(WildcardMsg())), Par())
+        )
+      )
+    val pattern: Match =
+      Match(
+        EVar(FreeVar(0)),
+        Seq(MatchCase(EList(Seq(EVar(FreeVar(0)), EVar(FreeVar(1)))),
+                      EVar(Wildcard(WildcardMsg()))),
+            MatchCase(EVar(Wildcard(WildcardMsg())), EVar(FreeVar(1))))
```

Comment:
New line after `Seq(` to match the indenting/newline structure of target would make this easier to mentally pattern match.